Contributing.md: Encourage discussion, improve commit message guide

paramat opened the pull request:
I have had some requests for an improved guide to writing commit messages.
I added a new step for contributing, to encourage discussion of a contribution before starting work on code, to save people's time, to help avoid wasted work, and reduce contributor frustration and disappointment.
Also fixed the number of hashes for certain title lines.

srifqi commented on CONTRIBUTING.md:
```diff
@@ -14,23 +14,31 @@ If you are planning to start some significant coding, you would benefit from ask
+        - Have no full stop at the end
+        - Not contain a Github issue or pull request number
+    - The second line should be empty
+    - The following lines should describe the commit, using a new line / paragraph for each point
```
Shouldn't there be no space around slash (/)?
`line/paragraph`

Ezhh replied:
I'd say drop "paragraph" as it's not needed.

paramat replied:
Ok.

nerzhul commented on CONTRIBUTING.md:
```diff
@@ -14,23 +14,31 @@ If you are planning to start some significant coding, you would benefit from ask
+        - Be a compact summary of the commit
+        - Have less than 70 characters
+        - Have no full stop at the end
+        - Not contain a Github issue or pull request number
```
i don't agree with that, it permits to have directly the issue or PR fixed, it's the most useful place.

paramat replied:
It seems better to have it in the later lines, a thread number is not descriptive and takes up precious room in the first line. Note, such an included number is not displayed as a link, and will be next to the link added by Github, so visually messy and a little confusing.

paramat replied:
But, not important, so ok.

nerzhul commented on CONTRIBUTING.md:
```diff
@@ -14,23 +14,31 @@ If you are planning to start some significant coding, you would benefit from ask
+        - Start with a capital letter
+        - Be a compact summary of the commit
+        - Have less than 70 characters
+        - Have no full stop at the end
```
We don't care about that

rubenwardy replied:
Nah, keep it

paramat replied:
Heh ok.

paramat replied:
Kept.

nerzhul commented on CONTRIBUTING.md:
```diff
@@ -14,23 +14,31 @@ If you are planning to start some significant coding, you would benefit from ask
+    - The first line should:
+        - Start with a capital letter
+        - Be a compact summary of the commit
+        - Have less than 70 characters
```
80 is the git limit. 70 is ridiculous

paramat replied:
I counted the number of characters before a '...' chop, see https://github.com/minetest/minetest/commit/cca925377b7929276de3bb04553ca6ee2012d35d the first line has 71 characters and is chopped after 69, so <70 seems correct.

paramat replied:
However i added the word 'Preferably ...'.

Ezhh commented on CONTRIBUTING.md:
```diff
@@ -14,23 +14,31 @@ If you are planning to start some significant coding, you would benefit from ask
 
-2. Start coding!
+2. Before you start coding, open an [issue at Github](https://github.com/minetest/minetest/issues) to discuss your intended contribution with the core developers, this may save time for you and others, and may avoid wasted work.
+
```
I'd advise making the issue sound more like a suggestion. ("open an issue..." is directly telling people to do it, even if the next bit implies it's optional.). Maybe "Consider opening an issue ... to discuss ..."? You should also look at the earlier part you haven't edited, which mentions discussing on IRC. This should all be combined into one part, with an issue or discussion via IRC suggested as alternative approaches. You might also add that it's worth searching existing issues?

New sentence should begin from "this may..."

Ezhh commented on CONTRIBUTING.md:
```diff
@@ -14,23 +14,31 @@ If you are planning to start some significant coding, you would benefit from ask
+        - Start with a capital letter
+        - Be a compact summary of the commit
+        - Have less than 70 characters
+        - Have no full stop at the end
```
But these lines here should all have them.

paramat replied:
Lol ok.

SmallJoker commented on CONTRIBUTING.md:
```diff
@@ -14,23 +14,31 @@ If you are planning to start some significant coding, you would benefit from ask
-        - Have a first line with less than *80 characters* and have a second line that is *empty*
-
-4. Once you are happy with your changes, submit a pull request.
+        - Be descriptive (e.g. no `Update init.lua` or `Fix a problem`)
```
IMO this line can be updated to so it provides more helpful, positive examples like:
`- Be descriptive (e.g. ^Fix segfault on world join^ or ^Implement entity inventory API^)`

paramat replied:
Ok.

paramat replied:
Looking for a good example.

nerzhul commented:
I agree with @SmallJoker . This PR is fully negative. if you want people to contribute, positive meaning is better

paramat commented on CONTRIBUTING.md:
```diff
@@ -14,23 +14,31 @@ If you are planning to start some significant coding, you would benefit from ask
 1. [Fork](https://help.github.com/articles/fork-a-repo/) the repository and [clone](https://help.github.com/articles/cloning-a-repository/) your fork.
 
-2. Start coding!
+2. Before you start coding, open an [issue at Github](https://github.com/minetest/minetest/issues) to discuss your intended contribution with the core developers, this may save time for you and others, and may avoid wasted work.
```
Yes this should be expressed as an option.

paramat commented:
'Fully negative'? That's ridiculous.
I do agree that opening an issue to discuss first should be expressed as optional and maybe a little less strictly, but 'this may save time for you and others, and may avoid wasted work.' is a positive intention.
The rest of this PR is improved commit message guidelines, nothing negative there.

paramat commented:
SmallJoker wrote:
> IMO this line can be updated to so it provides more helpful, positive examples

Nerzhul i agree with the above, the line referred to was not written by me.

paramat commented:
Updated.
SmallJoker yes i'm ok with improving that line, i just don't know what to put there. I guess it needs to be a short commit message that is also descriptive, i'll look through our commits for something, then will update again.

paramat commented:
Updated again, no longer WIP, added description of changes to first post.
A commit message example seems too long to include in the commit message guidelines and doesn't seem suitable for this document, so have referred to core dev commit messages as good examples.
